MDEV-34707: BUF_GET_RECOVER assertion failure on upgrade

dr-m opened the pull request:
- [x] *The Jira issue number for this PR is: [MDEV-34707](https://jira.mariadb.org/browse/MDEV-34707)*
## Description
`buf_page_get_gen()`: Relax the assertion once more. The LSN may grow by invoking `ibuf_upgrade()`, that is, when upgrading files where `innodb_change_buffering≠none` was used. The LSN may also have been recovered from a log that needs to be upgraded to the current format.
## Release Notes
N/A. This was a false alarm in a debug assertion whose intention is to catch the misuse of an internal API.
## How can this PR be tested?
1. Download the attachment from [MDEV-34707](https://jira.mariadb.org/browse/MDEV-34707)
2. `7z x dt.7z`
2. (Optional:) Start up and shut down MariaDB Server 10.5 or 10.6 with `--datadir` pointing to the `dt` subdirectory that was extracted in the previous step.
3. Start up MariaDB Server 11.x with `--datadir` pointing to the `dt `subdirectory.
## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature or a refactoring, and the PR is based against the latest MariaDB development branch.*
- [ ] *This is a bug fix, and the PR is based against the earliest maintained branch in which the bug can be reproduced.*
The earliest applicable branch is 11.1, but I think that this will be pushed after the pending quarterly releases, at which point the 11.2 branch would be the last maintained one.
## PR quality check
- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [ ] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=3451) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=3451) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=3451) it.</sub>

mariadb-DebarunBanerjee commented on storage/innobase/srv/srv0start.cc:
```diff
@@ -1182,6 +1185,10 @@ static tpool::task rollback_all_recovered_task(trx_rollback_all_recovered,
 					       nullptr,
 					       &rollback_all_recovered_group);
 
+#ifdef UNIV_DEBUG
+bool must_upgrade_ibuf;
+#endif
+
```
I could not understand this change. There is no reference to this global. There is one local variable (non-debug) exists by the same name and used. Something to remove ?

dr-m replied:
Right, this should not have been added. I added `ibuf_upgrade_was_needed` instead.

mariadb-DebarunBanerjee reviewed (approved):
Thanks for fixing the debug specific issue. Please go ahead.

We might want to eliminate the assert for checking LSN altogether. I checked the reasoning but It doesn't seem to be good validating against some constant value. Also, checking for LSN in buf_page_get_gen looks a bit misplaced.

dr-m commented:
> Thanks for fixing the debug specific issue. Please go ahead.
> 
> We might want to eliminate the assert for checking LSN altogether. I checked the reasoning but It doesn't seem to be good validating against some constant value. Also, checking for LSN in buf_page_get_gen looks a bit misplaced.

The purpose of that assertion is to ensure that `mode == BUF_GET_RECOVER` and `recv_sys.recover()` are not being misused. Very few page read operations may be invoked while the log-based recovery is in progress. Similarly, we want to ensure that any page reads during recovery will use `mode == BUF_GET_RECOVER`. The reason for this special logic is that 850d61736deba354377634cf344256ee850b93b7 simplified some code that had been added in 6697135c6d03935118c3dfa1c97faea7fa76afa6 to fix a recovery bug. Sometimes when `recv_sys.recover()` is invoked, recovery may not actually be running. The LSN assertions are related to that.

dr-m committed:
MDEV-34707: BUF_GET_RECOVER assertion failure on upgrade

buf_page_get_gen(): Relax the assertion once more.
The LSN may grow by invoking ibuf_upgrade(), that is,
when upgrading files where innodb_change_buffering!=none was used.
The LSN may also have been recovered from a log that needs
to be upgraded to the current format.
